LandSandBoat/server
# [JAs] Add functionality to WAR JA Restraint

**_I affirm:_**

## What does this pull request do?



- Creates new mod "ENHANCES_RESTRAINT" in status.lua and modifier.h
- Updates next available mod ID in modifier.h.
- Updates mods_by_id documentation with new mod.
- Enables Restraint in the ability.sql table.
- Adds the "ENHANCES RESTRAINT" mod to the Ravager's Muffler's +1/+2 and Boii Mufflers NQ/+1 in the item_mod.sql table.
- Sets up the Restraint effect .lua to handle the effect wearing off.
- Sets up job_utils/warrior.lua to add the Restraint status effect on job ability use.
- Adds Restraint handling while auto attacking with the Restraint status effect active to attack.cpp.
- Handling includes "Enhances Restraint" gear bonuses and "Restraint Effect" Job Points bonuses.
- Adds Restraint bonus and status effect power removal/reset after using a weaponskill to weaponskill_state.cpp.

**Notes**

Base equation has been established on BGWiki as:
`% Boost per Round = Floor( (Round Delay) / 385 )`

The final equation for this PR has been set up as follows:
`% Boost per Round = Floor( (Round Delay) / 385 ) * 1 + (Enhances Restraint / 100) * 1 + (Job Points / 100)`

Per both sources the equation only returns whole percentages of Weaponskill Damage bonus, so the minimum value received from this calculation will be 1.
The % Boost per Round is floored to the next lowest whole number but the remainder of the bonus is not lost and is recorded in the Restraint effect's Sub Power. This bonus is then applied to the next swing and maintains the benefit of using Job Points and higher bonus "Enhances Restraint" gear. Without tracking the remainder, all tiers of bonuses (Enhances, Job Point, and Enhances + Job Points) cap at a similar point and do not contribute to the calculation.

Per BGWiki the ability "is worth approximately 0.75 weapon skills without any enhancing gear" when using a 482 Delay weapon with 50% Haste. I have calculated this to mean that it will take the player 
`120s (+30% WSD & WS) + 120s (+30% WSD & WS) + 60s (+15% WSD & WS)`
for a total of 75% additional weapon skill damage in 5 minutes.

"Enhances Restraint" gear appears to directly double the base Boost per Round calculation, resulting in:
`60s (+30% WSD & WS) + 60s (+30% WSD & WS) + 60s (+30% WSD & WS) + 60s (+30% WSD & WS) + 60s (+30% WSD & WS)`
for a total of 150% additional weapon skill damage in 5 minutes.

The Job Point "Restraint Effect" then increases the Boost per Round value by 40%, further reducing the time to cap the effect.

**References**
https://www.bg-wiki.com/ffxi/Restraint
https://ffxiclopedia.fandom.com/wiki/Restraint

## Steps to test these changes



1. Upload the changes to your server.
2. Add `player:PrintToPlayer(string.format("WS Dmg First Hit: %i ", target:getMod(xi.mod.ALL_WSDMG_FIRST_HIT)), xi.msg.channel.SYSTEM_3)` to commands/getstats.lua.
3. Load into the game and go to your favorite testing mob.
4. Change to WAR using !changejob WAR 99.
5. Activate Restraint and begin attacking the mob.
6. Keep in mind that the effect will only update on the first swing of an attack round.
7. Use `!getstats <me>` to check your WSD stat as you attack.
8. Use a weapon skill and use `!getstats <me>` to check that effect reset to 0. Bonus may have increased due to a normal attack immediately after the weaponskill.
9. Use `!additem 11104`, `!additem 11204`, `!additem 27052`, or `!additem 27053` to give yourself the Ravager's/Boii Mufflers.
10. Equip the Hand armor and WSD bonus per hit should double.
11. Add Job Points to your character in the Mog House and max the "Restraint Effect" category.
12. Test the increases using `!getstats <me>`. The remainder will be catching most of this benefit and increasing the bonus of the next hit.
